ppy/osu
# Implement audio normalization using BASS

Closes #1600

Prereqs:
- https://github.com/ppy/osu-framework/pull/6233
- https://github.com/ppy/osu-framework/pull/6678
- https://github.com/ppy/osu-framework/pull/6281

## Audio Normalization
Audio Normalization is the process of determining the loudness of audio and standardizing (normalizing) it to a set level. If you normalize a number of audio files to the same level, they will all sound the same volume wise. Normalization is useful for making sure that loud songs get made quieter so that they aren't so loud, and quiet songs get made louder so that they aren't so quiet. This reduces the number of times you will need to change your volume, and yes, it means you can finally play The Big Black without having your eardrums reduced to shreds.

Here's a video showing what Audio Normalization looks like in osu!. In the video, all 3 of my volume sliders are set to max. The beatmap songs are being normalized to -14 LUFS. Try to focus on how every song sounds volume wise, you should find that they all sound about the same in terms of volume no matter the instruments in the song. I find that tech maps work the best when trying to tell what normalization does

https://github.com/ppy/osu/assets/69545310/847ea984-49bb-4010-b91c-9e74b8479fea

## Description of Approach
So, I decided to take a different approach than the old PR. The old PR tried implementing the algorithm manually which isn't good since we'd have to manually maintain it. This PR uses BASSLoud which is maintained by BASS and will get regular updates. I also noticed a lot of missing things in the old PR, such as recalculating the normalization values in the background. I've taken more or less the same structure as the old PR but improved upon it heavily.

When talking about approach in terms of normalization, this implementation calculates loudness using the [ITU-R BS.1770-4](https://web.archive.org/web/20231015224520/https://www.itu.int/dms_pubrec/itu-r/rec/bs/R-REC-BS.1770-4-201510-I!!PDF-E.pdf) standard, but normalizes to -14 LUFS instead. I've spent a couple hours playing the game on my branch to get a feel for the normalization, and I've found -14 LUFS to best meet my expectations. It makes the audio not too quiet, where the effects are really loud, but also loud enough that 100% volume actually feels like 100% volume. Additionally, -14 LUFS is what Youtube & Spotify normalize to, so I believe that it's probably the best level to normalize to when it comes to music. Whereas -23, as recommended by EBU, is best for movies and TV programmes.

My PR uses Integrated Loudness, meaning the average loudness across the entire song. This is different from more conventional methods of loudness detection, such as True Peak, which normalizes based on the peaks (parts which are loudest) in the song. Integrated Loudness is known for working better than True Peak since it takes into account how the song sounds like. Meaning sudden sounds in a song (say sfx), a sudden increase in vocal volume (such as in a chorus), or a sudden increase in the volume of an instrument (say at the end of a song as a finisher), will not be accounted for. These sudden sounds will not influence how the song is normalized, and won't be the reason for a relatively quiet but suddenly loud song to be made more quiet because of that sudden section. The old pr seemed to calculate and use a variety of methods, including True Peak, which was completely unnecessary.

## Description of Changes
This PR adds the ability for osu! to normalize audio tracks to a target level using BassLoud. This implementation goes through realm and tries to find maps which do not have loudness detection values stored, it then calculates these values via a background process. Once these values are calculated, they are stored and cached in Realm. These values are grabbed if they exist when a beatmap's audio track is loaded, if they don't exist, the old global audio reduction is applied (0.8). The values are used to create a Volume Fx filter for the track to normalize it.

## Changes to Realm
Since we're now tracking Audio Normalization values in Realm, we need to bump the version otherwise a migration error is thrown. Apart from creating object mappings, the only change to realm here is a bump to the version.

## Testing
Testing this is a bit difficult since there are many changes in many areas. It should get exponentially easier as things begin to get merged.

1. Clone the [osu branch in this PR](https://github.com/smallketchup82/osu/tree/audio-normalization), my [ManagedBass branch](https://github.com/smallketchup82/ManagedBass/tree/peppy-fork-smallketchup-edit), and my [osu-framework branch](https://github.com/smallketchup82/osu-framework/tree/bass-loudness-normalization).
2. In `osu-framework`, pack `NativeLibs` into a nupkg, then add a nuget source for it. Downgrade the `NativeLibs` version that `osu-framework` is using to `1.0.0`.
3. Checkout hwsmm's [osu-framework branch](https://github.com/hwsmm/osu-framework/tree/audio-normali)
4. Run the following commands in the osu-framework directory (assuming you cloned everything in the same directory, if not you'll have to adjust these paths yourself):
```shell
dotnet remove osu.Framework/osu.Framework.csproj ppy.ManagedBass
dotnet add osu.Framework/osu.Framework.csproj reference ../ManagedBass/src/Bass/Portable/Bass.csproj
dotnet add osu.Framework/osu.Framework.csproj reference ../ManagedBass/src/AddOns/BassLoud/BassLoud.csproj
dotnet restore osu.Framework
```
6. Go back to osu! lazer (`osu`) and run `UseLocalFramework.sh`. For good luck, run `dotnet restore osu.Desktop` just to make sure everything is correctly restored
7. If all is well, you should be able to run `dotnet run` (or ide equivalent) in `osu.Desktop` with no problems.

Feel free to ping me in the `#osu` channel of the osu!dev discord if you need help with this.

## Final remarks (notes before reviewing)
Please read this section before reviewing
- **Please backup your realm file before testing!**
- Please go through this commit by commit. I leave a lot of useful information in my commit descriptions. Start at 7a3ccf3
	- Sorry about the commits before that one, I tend not to think about variable names in rough drafts and poc code
- I intentionally left out tests since this isn't really something you can just slap in a test. It works and tests best when using the full client and playing on the branch for a while.

Developed with help from @hwsmm (thanks!!!)
